src/main.rs

```diff
@@ -269,30 +297,19 @@ impl App {
     }
 
     fn next(&self, args: &ArgMatches) -> Result<()> {
-        let next_handle = self.handle_from_args(args).or_else(|_| {
-            let reverse = args.get_flag("reverse");
-            let fastest = args.get_flag("fastest");
-            let shortest = args.get_flag("shortest");
-            if reverse || fastest || shortest {
-                let max_attemps = 100;
-                for _i in 0..max_attemps {
-                    let handle = self.random_handle()?;
-                    let clash = self.read_clash(&handle)?;
-                    if (!reverse || clash.is_reverse())
-                        && (!fastest || clash.is_fastest())
-                        && (!shortest || clash.is_shortest())
-                    {
-                        return Ok(handle)
-                    }
+        let next_handle = match args.get_one::<PublicHandle>("PUBLIC_HANDLE") {
+            Some(h) => h.to_owned(),
+            None => {
+                let fastest = args.get_flag("fastest");
+                let shortest = args.get_flag("shortest");
+                let reverse = args.get_flag("reverse");
+                if reverse || fastest || shortest {
+                    self.random_handle_with_modes(fastest, shortest, reverse)?
+                } else {
+                    self.random_handle()?
                 }
-                Err(anyhow!(format!(
-                    "Failed to find a clash with the required modes after {} attempts.",
-                    max_attemps
-                )))
-            } else {
-                self.random_handle()
             }
-        })?;
+        };
         println!("Changed clash to https://codingame.com/contribute/view/{}", next_handle);
         println!(" Local file: {}/{}.json", &self.clash_dir.to_str().unwrap(), next_handle);
```

Comment:
Currently the `clash next` command won't throw you an error if you put random 0-9a-f gibberish as a HANDLE. It will only be when you try `clash run` that an error will be sent to you as that file doesn't exist.

I tried this that seems to work:

```
        let file_path = format!("{}/{}.json", &self.clash_dir.to_str().unwrap(), next_handle);
        if !std::fs::metadata(&file_path).is_ok() {
            return Err(anyhow!(format!("The file {} does not exist", &file_path)));
        }
        println!("Changed clash to https://codingame.com/contribute/view/{}", next_handle);
        println!(" Local file: {}", file_path);
```

![image](https://github.com/Andriamanitra/clash/assets/80173028/5807d7af-9ec6-4d2b-8704-3af198f425af)

Also, is the space before "Local file" intended?

Reply:
> Currently the clash next command won't throw you an error if you put random 0-9a-f gibberish as a HANDLE. 

I think `clash next` not erroring is kind of harmless, you'll still get an error if you try to actually read the clash file. If we want to make it an error the easiest way to make sure error message is consistent would be to read the clash with `self.read_clash(&next_handle)?` (even though it's unnecessary for the functionality of the `clash next` command, but that would also ensure an error if the clash is in invalid format or something).

> Also, is the space before "Local file" intended?

Yes, it's actually the other print missing a space before it that was not intended originally. :smile:

The spaces make it looks way less messy with my terminal setup (screenshot below), but both of the prints ("Changed clash.." and " Local file: ...") were put there very early on pretty much just for debugging purposes, I'm not sure if they should even stay?

![image](https://github.com/Andriamanitra/clash/assets/10672443/e2048009-5cb3-4916-8877-853d1b873b15)

Reply:
You could make a simple get_json_path function that is just:

```
   fn () -> String {
        let file_path = format!("{}/{}.json", &self.clash_dir.to_str().unwrap(), next_handle);
        if !std::fs::metadata(&file_path).is_ok() {
            return Err(anyhow!(format!("The file {} does not exist", &file_path)));
        };
        file_path
   }
```

And jam it in both `clash next` and `read_clash` for a consistent message. You don't need to parse the clash, just to check that the file exists.

About the spaces, I would make it consistent, as in either one space at each line or none at each line. One and then none feels off. Maybe one is better?

And if you meant removing the messages altogether, I think it's fine for now? The local file one is especially useful for setting up things at the start.